docs/design/sql/16-nested-fields-access.md

```diff
@@ -0,0 +1,359 @@
+consider JDBC to be the primary Java API. This is also what PG and Oracle do
+(TODO confirm Oracle). The user can reasonably only use `toString` on the value,
+which is what JDBC UI clients do, the `toString` will contain comma-separated
+list of all fields.
+
+The `ROW` type supports comparison operators that is different from how the
+comparison operator works with `OBJECT` type. It compares all field one by one,
+from left to right (this is important for operators other than `=`).
+
+The `ROW` type can also be assigned to another `ROW` type, provided they have
+the same number and types of fields. This is important with `ROW` literals, they
+can be used to create UDT values:
+
+```sql
+UPDATE m
+SET address=CAST(('street', 'city', 12345, 'Slovakia') AS AddressType
+WHERE ...;
+```
+
+### Semantics of comparison operators
+
+Comparison operators are: `=`, `!=`, `<`, `>`, `<=`, `>=`. `IS [NOT] DISTINCT
+FROM` isn't currently supported.
+
+When one operand of a comparison operator is ROW, the other operand must:
+- have a ROW type
+- have the same number of fields
+
+For individual fields of the ROW, we apply normal rules for comparison operator
+recursively. That is, `(value1, value2) = (value3, value4)` is equivalent to
+`value1 = value3 AND value2 = value4`, including the type conversion this
+involves.
+
+For example: 
+
+```
+(INT, INT) <cmp> (BIGINT, INT)   # ok, INT converted to BIGINT
+(INT, INT) <cmp> OBJECT  # reject ROW and OBJECT comparison
+(INT) <cmp> (VARCHAR) # ok, VARCHAR onverted to INT
+(INT) <cmp> (INT, INT)  # reject - different number of fields
+(INT, (INT, BIGINT)) <cmp> (TINYINT, (BIGINT, INT))  # ok, same rules for nested ROW
+```
+
+The comparison will be done from left to right. For evaluation, initial equal
+fields are ignored, and the first non-equal field determines the result. If all
+fields are equal, any field determines the result.
+
+### Issues with `CAST(field AS ROW)`
+
+For some reason we weren't able to implement this (TODO ivan).
+
+## Information schema
+
+We need to provide these new tables:
+- `USER_DEFINED_TYPES`
+- `ATTRIBUTES`
+
+We should provide `INFORMATION_SCHEMA.USER_DEFINED_TYPES` view, which is SQL
+standard. I couldn't find an information schema table to list type fields.
+
+# Implementation details
+
+## CREATE/DROP TYPE statement
+
+Syntax:
+```sql
+CREATE [OR REPLACE] TYPE [IF NOT EXISTS] <typeName> [<columnList>] 
+OPTIONS ('format'='<typeFormat>',
+  <...other options relevant for the chosen format...>)
+
+<columnList>:
+(
+    columnName ColumnType,
+    [<... more columns ...>]
+)
+
+DROP TYPE [IF EXISTS] <typeName>
+```
+
+## Catalog storage
+
+Types are stored in `__sql.catalog` map, along with mappings and views. This
+ensures name uniqueness.
+
+As with other SQL metadata, we don't enforce mutual dependencies. For example, a
+type can be dropped even if it is used in some mapping - when this happens, the
+mapping then becomes invalid. (TODO test this)
```

Comment:
In current implementation types are preserved inside of mappings (and mappings are not invalided upon Type change). Do we need this mechanism actually? I don't foresee types being changed much after creation, but maybe it would be more correct to break the mappings deliberately in this case (and also change them if Type changes).

Reply:
https://github.com/hazelcast/hazelcast/issues/22129